Best practice of replacing nic (#511)
harvester/docs: docs/networking/best-practice.md

---
sidebar_position: 6
sidebar_label: Harvester Network Best Practice
title: "Harvester Network Best Practice"
keywords:
- Harvester
- Networking
- Best Practice
---

<head>
  <link rel="canonical" href="https://docs.harvesterhci.io/v1.1/networking/best-practice"/>
</head>

## Replace Ethernet Nics

You may plan to replace the ethernet nics of a bare-metal node in the running Harvester cluster due to the following reasons:

- Nics are broken
- Nics are with insufficent hardware capacity
- Nics are not with expected features

> You may follow the below steps and run them in each node step by step.

Author: This md is for future extension
`# best practice`
`## Replace Ethernet Nics`
`## e.g. Utilize the IPPool for LB`...

all the contents in this PR need to be under `## Replace Ethernet Nics`

### Validate the new Nics

1. Confirm with SUSE/Harvester if the new Nics are certificated by the related Harvester version.

1. Test the new nics in non-production environment.

### Pre-check

1. Check all VMs from the Harvester [VM page](../vm/access-to-the-vm.md#access-with-the-harvester-ui), they should be either running or stopped.

1. Check all Longhorn volumes are healthy from the [embedded Longhorn dashboard](../troubleshooting/harvester.md#access-embedded-rancher-and-longhorn-dashboards). It also displays those none-VM related volumes.

1. (Optional) Generate a [support bundle](../troubleshooting/harvester.md#generate-a-support-bundle) for comparison with the new node later.

### Collect Information

> Before any action is taken, it is important to collect the current network information and status.

Reviewer: All items are pre-procedure checks. Placing the first two items in a separate chapter does not communicate a difference in "timing".

        STARTMODE='onboot'
        BONDING_MASTER='yes'
        BOOTPROTO='none'
        POST_UP_SCRIPT="wicked:setup_bond.sh"

        BONDING_SLAVE_0='ens3'

        BONDING_MODULE_OPTS='miimon=100 mode=active-backup '

        DHCLIENT_SET_DEFAULT_ROUTE='no'

      encoding: ""
      ownerstring: ""

    - path: /etc/sysconfig/network/ifcfg-ens3
      permissions: 384
      owner: 0
      group: 0
      content: |
        STARTMODE='hotplug'
        BOOTPROTO='none'
      encoding: ""
      ownerstring: ""
    ```

- ip link

    `ip link` displays all the physical nics, their master if exists, and the `state`.

    Example:

    ```
    ip link | grep master -1

    2: ens3: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast master mgmt-bo state UP mode DEFAULT group default qlen 1000
        link/ether 52:54:00:03:3a:e4 brd ff:ff:ff:ff:ff:ff
    --

    4: mgmt-bo: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1500 qdisc noqueue master mgmt-br state UP mode DEFAULT group default qlen 1000
        link/ether 52:54:00:03:3a:e4 brd ff:ff:ff:ff:ff:ff

    ```

- lspci

    `lspci` shows all pci devices in summary, it is easy to identify the network nics.

>     Example:

Author: This doc, is more for IT/network adminstrators to follow to do things step by step, it is more technical, please keep original format.

>     Example:

Reviewer: I disagree. The section title is "Collect Information" and the first item in your draft is a type of information. This is not a procedure because you used an unordered list. If you wanted users to follow steps, you should have used an ordered list. Lastly, I did not remove any technical information. I simply made it easier to understand.

You can reject the suggestion if you want.

- Wait for everything to get ready, repeat the steps in [pre-check](#pre-check).

- Stop a VM manually when:

    1. The VM fails to migrate
    1. The VM has some selectors which cause it has no other nodes to migrate
    1. The VM has special hardwares like pci-passthrough or vgpu and fails to migrate

### Replace the Nics

1. Shutdown the node
1. Replace the NICs
1. Restart the node

### Check the new Nics Information

Repeat the operations in [collect information](#collect-information) and check them.

If anything is abnormal, please hold on to this step and double-check with Harvester. Generate a [support bundle](../troubleshooting/harvester.md#generate-a-support-bundle) for troubleshooting.

### Disable the Maintenance Mode

1. Wait the [node](../host/host.md) is back to cluster.

1. Disable the maintenance mode and continue wait until everything is ready.

1. (Optional) Start the VMs which were manually stopped.

1. (Optional) Manually [migrate](../vm/live-migration.md#starting-a-migration) some VMs to this node.

### Troubleshooting

> Harvester has a couple of network related pods and CRDs. It is very helpful to check the logs of pods, the status of CRD objects.

Reviewer: ```suggestion
Harvester uses multiple network-related pods and CRDs. When troubleshooting, check the pod logs and the status of CRD objects.
```

    Jan 08 00:34:48 localhost kernel: i40e 0000:5d:00.1 eth1: NIC Link is Up, 10 Gbps Full Duplex, Flow Control: None
    Jan 08 00:34:48 localhost kernel: i40e 0000:af:00.0 eth4: NIC Link is Up, 10 Gbps Full Duplex, Flow Control: None
    Jan 08 00:34:49 localhost kernel: i40e 0000:af:00.1 eth2: NIC Link is Up, 10 Gbps Full Duplex, Flow Control: None
    ```

### Enable Maintenance Mode

1. (Optional) Stop VMs that cannot or must not be migrated.

1. [Enable maintenance mode](../host/host.md#node-maintenance) on the target node to automatically migrate all VMs to other nodes.

- Wait for everything to get ready, repeat the steps in [pre-check](#pre-check).

- Stop a VM manually when:

    - The VM fails to migrate

    - The VM has some selectors which cause it has no other nodes to migrate

    - The VM has special hardwares like pci-passthrough or vgpu and fails to migrate

### (Optional) Update the VM Network

:::note

When the new Nics will be placed in different physical slots or those nics will have different uplink parameters, this step is mandatory.

:::

1. Check the node.

>     For a Harvester cluster node, when it is belonging to a VM Network (VlanConfig), the node object has a label with key `network.harvesterhci.io/vlanconfig`.

Reviewer 2: I remember that VM Networks maps to NADs. It's Network Configs that map to VlanConfig. We might need to update the term here.

    The running replicas on the node will be stopped at this stage. They will be shown as Failed.

    The engine processes on the node will be migrated with the Pod to other nodes.

    After the drain is completed, there should be no engine or replica process running on this node.

1. Replenish replicas.

    After a node is shutdown, Longhorn will not start rebuilding the replicas on other nodes until `replica-replenishment-wait-interval` (default 600 s) is exceeded. When the node comes back online before the wait-interval, Longhorn can reuse the replicas. If not, Longhorn will rebuild the replicas on another node.

    In a system maintenance, you may already know the node will be down for long time, then you can adjust the [replica-replenishment-wait-interval](https://longhorn.io/docs/1.4.3/references/settings/#replica-replenishment-wait-interval) from the from the [embedded Longhorn dashboard](../troubleshooting/harvester.md#access-embedded-rancher-and-longhorn-dashboards) to guide Longhorn to rebuild the replicas quickly.

    Harvester v1.3.0 has Longhorn v1.6.0 embedded. Harvester v1.2.1 has Longhorn v1.4.3 embedded.

### Replace the Nics

> 1. Shutdown the node.

Reviewer: ```suggestion
1. Shut the node down.
```

1. Replace the NICs.

1. Restart the node.

1. [Collect information](#collect-information) about the current network configuration and status.

If you observe any abnormalities, generate a [support bundle](../troubleshooting/harvester.md#generate-a-support-bundle) for troubleshooting purposes.

### (Optional) Re-update the VM Network

:::note

When the new Nics are placed in different slots, this step is mandatory.

:::

1. Add this node to the VM network.

>     You need to create a new VM Network (VlanConfig) or change the existing VM Network (VlanConfig) to include this Node.

Reviewer 2: Ditto

> 1. Add this node to the VM network.
> 
>     You need to create a new VM Network (VlanConfig) or change the existing VM Network (VlanConfig) to include this Node.

Reviewer: ```suggestion
1. Add the node to the VM network.

    You must create a new VM network (VlanConfig) or change the VM network to include this node.
```

> 1. Wait the node.
> 
>     Wait and check the `Node` object, until it's label `network.harvesterhci.io/vlanconfig` has this VlanConfig.

Reviewer: ```suggestion
1. Check the `Node` object.

    The label `network.harvesterhci.io/vlanconfig` reflects the specific VM network used.
```

> 1. Wait the VlanStatus.
> 
>     Wait and check the related `VlanStatus` object until its `ready` condition is "True".

Reviewer: ```suggestion
1. Check the `VlanStatus` object.

    The status of the `VlanStatus` object's `ready` condition changes to `"True"`.
```
